trillian/migrillian/core/controller.go

```diff
@@ -50,19 +52,59 @@ type Controller struct {
 	batches  chan scanner.EntryBatch
 	ctClient *client.LogClient
 	plClient *PreorderedLogClient
+	ef       election.Factory
 }
 
-// NewController creates a Controller configured by the passed in options.
-func NewController(opts Options, ctClient *client.LogClient, plClient *PreorderedLogClient) *Controller {
-	bufferSize := opts.Submitters * opts.BatchesPerSubmitter
-	batches := make(chan scanner.EntryBatch, bufferSize)
-	return &Controller{opts: opts, batches: batches, ctClient: ctClient, plClient: plClient}
+// NewController creates a Controller configured by the passed in options, CT
+// and Trillian clients, and a (possibly nil) master election factory.
+func NewController(opts Options, ctClient *client.LogClient,
+	plClient *PreorderedLogClient, ef election.Factory,
+) *Controller {
+	return &Controller{opts: opts, ctClient: ctClient, plClient: plClient, ef: ef}
+}
+
+// RunWithElection is a master-elected version of Run method. It executes Run
+// whenever this instance captures mastership of the tree ID. As soon as the
+// instance stops being the master, Run is canceled. The method returns if an
+// error occurs, the passed in context is canceled, or fetching is done (in
+// non-Continuous mode).
+func (c *Controller) RunWithElection(ctx context.Context) error {
+	logID := c.plClient.tree.TreeId
+
+	el, err := c.ef.NewElection(ctx, logID)
+	if err != nil {
+		return err
+	}
+	defer func(ctx context.Context) {
+		if err := el.Close(ctx); err != nil {
+			glog.Warningf("Election.Close(): %v", err)
```

Comment:
self-nit: Could add a logID to the log message, as this code will eventually be multi-tenant.